modules/xfeatures2d/src/beblid.cpp

```diff
@@ -0,0 +1,602 @@
+    int descriptorSize() const CV_OVERRIDE { return int(wl_params_.size() / 8); }
+
+    // returns the descriptor type
+    int descriptorType() const CV_OVERRIDE { return CV_8UC1; }
+
+    // returns the default norm type
+    int defaultNorm() const CV_OVERRIDE { return cv::NORM_HAMMING;  }
+
+    // compute descriptors given keypoints
+    void compute(InputArray image, vector<KeyPoint> &keypoints, OutputArray descriptors) CV_OVERRIDE;
+
+private:
+    std::vector<ABWLParams> wl_params_;
+    float scale_factor_;
+    cv::Size patch_size_;
+
+    void computeBEBLID(const cv::Mat &integralImg,
+                       const std::vector<cv::KeyPoint> &keypoints,
+                       cv::Mat &descriptors);
+}; // END BEBLID_Impl CLASS
+
+/**
+ * @brief Function that determines if a keypoint is close to the image border.
+ * @param kp The detected keypoint
+ * @param imgSize The size of the image
+ * @param patchSize The size of the normalized patch where the measurement functions were learnt.
+ * @param scaleFactor A scale factor that magnifies the measurement functions w.r.t. the keypoint.
+ * @return true if the keypoint is in the border, false otherwise
+ */
+inline bool isKeypointInTheBorder(const cv::KeyPoint &kp,
+                                  const cv::Size &imgSize,
+                                  const cv::Size &patchSize = {32, 32},
+                                  float scaleFactor = 1)
+{
+    // This would be the correct measure but since we will compare with half of the size, use this as border size
+    float s = scaleFactor * kp.size / (patchSize.width + patchSize.height);
+    cv::Size2f border(patchSize.width * s * CV_BEBLID_EXTRA_RATIO_MARGIN,
+                      patchSize.height * s * CV_BEBLID_EXTRA_RATIO_MARGIN);
+
+    if (kp.pt.x < border.width || kp.pt.x + border.width >= imgSize.width) return true;
+    if (kp.pt.y < border.height || kp.pt.y + border.height >= imgSize.height) return true;
+    return false;
+}
+
+/**
+ * @brief Rectifies the coordinates of the measurement functions that conform the descriptor
+ * with the keypoint location parameters.
+ * @param wlPatchParams The input weak learner parameters learnt for the normalized patch
+ * @param wlImageParams The output  weak learner parameters adapted to the keypoint location
+ * @param kp The keypoint defining the offset, rotation and scale to be applied
+ * @param scaleFactor A scale factor that magnifies the measurement functions w.r.t. the keypoint.
+ * @param patchSize The size of the normalized patch where the measurement functions were learnt.
+ */
+inline void rectifyABWL(const std::vector<ABWLParams> &wlPatchParams,
+                        std::vector<ABWLParams> &wlImageParams,
+                        const cv::KeyPoint &kp,
+                        float scaleFactor = 1,
+                        const cv::Size &patchSize = cv::Size(32, 32))
+{
+    float m00, m01, m02, m10, m11, m12;
+    float s, cosine, sine;
+    uint32_t i;
```

Comment:
Consider placing vars declaration near its usage.

> uint32_t i

`for (size_t i = 0; i < wlPatchParams.size(); i++)` ?

Reply:
done